Sources/SCTE35/ModelObjects/SpliceDescriptors/SegmentationInfo.swift

```diff
@@ -343,75 +349,214 @@ public struct SegmentationUPID {
+        guard relevantBits.count >= 16 else { return nil }
+
+        var upids = [SegmentationUPID]()
+
+        var bits = relevantBits
+        while bits.count > 0 {
+            let upidTypeRange = 0..<8
+            let upidType = BitConverter.integer(fromBits: Array<Bit>(bits[upidTypeRange]))
+            let upidLengthRange = 8..<16
+            let upidLengthInBytes = BitConverter.integer(fromBits: Array<Bit>(bits[upidLengthRange]))
+            let startIndex = 16
+            let endIndex = 16 + (upidLengthInBytes * 8)
+            guard bits.count >= endIndex else { return nil }
+            let segmentationBitsRange = startIndex..<endIndex
+            let upidBits = Array<Bit>(bits[segmentationBitsRange])
+            guard let segmentationUPID = SegmentationUPID(type: upidType, length: upidLengthInBytes, relevantBits: upidBits) else {
+                return nil
+            }
+            upids.append(segmentationUPID)
+            bits.removeSubrange(0..<endIndex)
+        }
+
+        return upids
+    }
+}
+
+extension SegmentationUPID: BitEncodable {
+    func encode() throws -> [Bit] {
+        var bits = [Bit]()
+        bits.append(contentsOf: BitConverter.bits(from: type))
+        bits.append(contentsOf: BitConverter.bits(from: length))
+
+        switch info {
+        case .none:
+            break
+        case .userDefined(let value):
+            let userDefinedBits = BitConverter.bits(fromData: value)
+            bits.append(contentsOf: userDefinedBits)
+
+        case .ISCI(let value), .AdID(let value), .UMID(let value), .VISAN(let value),
+                .ISAN(let value), .TID(let value), .TI(let value), .ADI(let value),
+                .EIDR(let value), .MPU(let value), .ADS(let value), .URI(let value), .UUID(let value):
+            guard let stringData = value.data(using: .utf8) else { throw SCTE35ParsingError.invalidStringAsUtf8 }
+            let stringBits = BitConverter.bits(fromData: stringData)
+            bits.append(contentsOf: stringBits)
+
+        case .ATSC(let atsc):
+            let atscBits = try atsc.encode()
+            bits.append(contentsOf: atscBits)
+
+        case .MID(let upids):
+            let midBits = try encodeMultipleUpids(upids)
+            bits.append(contentsOf: midBits)
+
+        }
+
+        return bits
+    }
+
+    private func encodeMultipleUpids(_ upids: [SegmentationUPID]) throws -> [Bit] {
+        var allUpidBits = [Bit]()
+        for upid in upids {
+            let multiUpidBits = try upid.encode()
+            allUpidBits.append(contentsOf: multiUpidBits)
+        }
+        return allUpidBits
+    }
 }
 
-/// Additional Information that may be associated with Segmentation UPID
-public struct SegmentationUPIDInformation {
-    /// These have been tested against several, but not all of SegmentationUPID.type
-    /// Feel free to fork this repo and add additional decoding
-    public var string: String?
-    public var array: [SegmentationUPID]?
```

Comment:
I modified this for a few reasons:
1. This type will only ever have one of these two properties defined based on the parsed UPID type, thus one of these properties would always be nil and thus a little wasteful to have defined in memory.
2. We needed to support additional types such as `ATSCContentIdentifier` and `Data` so it felt silly to add 2 more optional properties of these types.
3. Enums with associated values are well suited to represent one of several possible UPID types that are represented in different ways.